Mage.Sets/src/mage/cards/a/AgencyOutfitter.java

```diff
@@ -0,0 +1,127 @@
+package mage.cards.a;
+
+import mage.MageInt;
+import mage.abilities.Ability;
+import mage.abilities.common.EntersBattlefieldTriggeredAbility;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.keyword.FlyingAbility;
+import mage.cards.*;
+import mage.constants.CardType;
+import mage.constants.Outcome;
+import mage.constants.SubType;
+import mage.constants.Zone;
+import mage.filter.FilterCard;
+import mage.filter.predicate.mageobject.NamePredicate;
+import mage.game.Game;
+import mage.players.Player;
+import mage.target.common.TargetCardInHand;
+import mage.target.common.TargetCardInLibrary;
+import mage.target.common.TargetCardInYourGraveyard;
+
+import java.util.AbstractMap;
+import java.util.UUID;
+import java.util.function.Function;
+
+/**
+ *
+ * @author notgreat
+ */
+public final class AgencyOutfitter extends CardImpl {
+
+    public AgencyOutfitter(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{4}{U}{U}");
+        this.subtype.add(SubType.SPHINX);
+        this.subtype.add(SubType.DETECTIVE);
+        this.power = new MageInt(4);
+        this.toughness = new MageInt(3);
+
+        // Flying
+        this.addAbility(FlyingAbility.getInstance());
+
+        // When Agency Outfitter enters the battlefield, you may search your graveyard, hand, and/or library for a card named Magnifying Glass and/or a card named Thinking Cap and put them onto the battlefield. If you search your library this way, shuffle.
+        this.addAbility(new EntersBattlefieldTriggeredAbility(new AgencyOutfitterEffect(), true));
+    }
+
+    private AgencyOutfitter(final AgencyOutfitter card) {
+        super(card);
+    }
+
+    @Override
+    public AgencyOutfitter copy() {
+        return new AgencyOutfitter(this);
+    }
+}
+
+//Based on Boonweaver Giant / Dark Supplicant / Gate to the Afterlife / Mishra, Artificer Prodigy
+class AgencyOutfitterEffect extends OneShotEffect {
+
+    AgencyOutfitterEffect() {
+        super(Outcome.UnboostCreature);
+        this.staticText = "you may search your graveyard, hand, and/or library for a card named Magnifying Glass and/or a card named Thinking Cap and put them onto the battlefield. If you search your library this way, shuffle.";
+    }
+
+    private AgencyOutfitterEffect(final AgencyOutfitterEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public AgencyOutfitterEffect copy() {
+        return new AgencyOutfitterEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+
+        //Uses AbstractMap.SimpleImmutableEntry as a generic Pair class
+        //Create a local lambda function to reduce code duplication
+        Function<String, AbstractMap.SimpleImmutableEntry<Card, Boolean>> searchEverywhereForCard = (name) -> {
```

Comment:
Do not use non standard constructions or types

Reply:
The `AbstractMap.SimpleImmutableEntry` is used as a pair type in a couple other cards. Using lambda functions is pretty common (though more frequently in streams). If you'd really prefer to have the code just copy&pasted I can, but that's a lot of highly repetitive code.

Reply:
AbstractMap used in 3 cards only. Must be replaced by normal map.

Same for “function<“ - do not see any usages of it. You can replace it by normal method with params.

Reply:
It's not a map, it's a map Entry (equivalent to javafx.util.Pair, but we aren't using javafx here). I could make a whole new class to store those two values, 

Yeah ok a private function works well enough, I'll use that.

Reply:
I don't understand why any cards use `AbstractMap.SimpleImmutableEntry`.